File: data/readable_format/fortran/file2062.f90
! combined.f90
SUBROUTINE DLAPMT(FORWRD, M, N, X, LDX, K)
  LOGICAL, INTENT(IN) :: FORWRD
  INTEGER, INTENT(IN) :: LDX, M, N
  INTEGER, INTENT(INOUT) :: K(LDX)
  DOUBLE PRECISION, INTENT(INOUT) :: X(LDX, *)

  INTEGER :: I, II, IN, J
  DOUBLE PRECISION :: TEMP

  IF (N.LE.1) RETURN

  DO I = 1, N
     K(I) = -K(I)
  END DO

  IF (FORWRD) THEN
     DO I = 1, N
        IF (K(I) .GT. 0) CYCLE

        J = I
        K(J) = -K(J)
        IN = K(J)

        DO
           IF (K(IN) .GT. 0) EXIT

           DO II = 1, M
              TEMP = X(II, J)
              X(II, J) = X(II, IN)
              X(II, IN) = TEMP
           END DO

           K(IN) = -K(IN)
           J = IN
           IN = K(IN)
        END DO
     END DO
  ELSE
     DO I = 1, N
        IF (K(I) .GT. 0) CYCLE

        K(I) = -K(I)
        J = K(I)

        DO
           IF (J .EQ. I) EXIT

           DO II = 1, M
              TEMP = X(II, I)
              X(II, I) = X(II, J)
              X(II, J) = TEMP
           END DO

           K(J) = -K(J)
           J = K(J)
        END DO
     END DO
  END IF
END SUBROUTINE DLAPMT

PROGRAM testDLAPMT
  IMPLICIT NONE
  INTEGER, PARAMETER :: dp = KIND(1.0D0)
  INTEGER :: M, N, LDX
  LOGICAL :: FORWRD
  INTEGER, ALLOCATABLE :: K(:)
  DOUBLE PRECISION, ALLOCATABLE :: X(:,:)
  INTEGER :: i, j

  ! Test parameters
  M = 2
  N = 3
  LDX = M
  ALLOCATE(K(N), X(LDX,N))

  ! Initialize data
  K = (/ 3, 1, 2 /)
  X = RESHAPE((/ 1.0_dp, 4.0_dp, 2.0_dp, 5.0_dp, 3.0_dp, 6.0_dp /), SHAPE(X))
  FORWRD = .TRUE.

  ! Call the subroutine
  CALL DLAPMT(FORWRD, M, N, X, LDX, K)

  ! Print the result
  PRINT *, "Result:"
  DO i = 1, M
      DO j = 1, N
          PRINT *, X(i, j)
      END DO
  END DO
END PROGRAM testDLAPMT

File: data/readable_format/cpp/file2062.cpp
#include <iostream>
#include <vector>

// Define the DLAPMT function within the same file
void DLAPMT(bool FORWRD, int M, int N, std::vector<std::vector<double>>& X, std::vector<int>& K) {
    int I, IN, J, II;
    double TEMP;

    if (N <= 1) return;

    for (I = 1; I <= N; ++I) {
        K[I - 1] = -K[I - 1]; // Adjust index for 0-based
    }

    if (FORWRD) {
        for (I = 1; I <= N; ++I) {
            if (K[I - 1] > 0) continue;

            J = I;
            K[J - 1] = -K[J - 1]; // Adjust index for 0-based
            IN = K[J - 1]; // Adjust index for 0-based

            while (true) {
                if (K[IN - 1] > 0) break;

                for (II = 1; II <= M; ++II) {
                    TEMP = X[II - 1][J - 1]; // Adjust index for 0-based
                    X[II - 1][J - 1] = X[II - 1][IN - 1]; // Adjust index for 0-based
                    X[II - 1][IN - 1] = TEMP; // Adjust index for 0-based
                }

                K[IN - 1] = -K[IN - 1]; // Adjust index for 0-based
                J = IN;
                IN = K[IN - 1]; // Adjust index for 0-based
            }
        }
    } else {
        for (I = 1; I <= N; ++I) {
            if (K[I - 1] > 0) continue;

            K[I - 1] = -K[I - 1]; // Adjust index for 0-based
            J = K[I - 1]; // Adjust index for 0-based

            while (true) {
                if (J == I) break;

                for (II = 1; II <= M; ++II) {
                    TEMP = X[II - 1][I - 1]; // Adjust index for 0-based
                    X[II - 1][I - 1] = X[II - 1][J - 1]; // Adjust index for 0-based
                    X[II - 1][J - 1] = TEMP; // Adjust index for 0-based
                }

                K[J - 1] = -K[J - 1]; // Adjust index for 0-based
                J = K[J - 1]; // Adjust index for 0-based
            }
        }
    }
}

// Main function for testing
int main() {
    int M = 2, N = 3;
    std::vector<int> K = {3, 1, 2}; // Adjust for 1-based indexing
    std::vector<std::vector<double>> X = {{1.0, 2.0, 3.0}, {4.0, 5.0, 6.0}};
    bool FORWRD = true;

    DLAPMT(FORWRD, M, N, X, K);

    std::cout << "Result:" << std::endl;
    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            std::cout << X[i][j] << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}